src/hotspot/share/oops/compressedOops.hpp

```diff
@@ -125,6 +128,19 @@ class CompressedOops : public AllStatic {
   static oop decode(oop v)                      { return v; }
   static narrowOop encode_not_null(narrowOop v) { return v; }
   static narrowOop encode(narrowOop v)          { return v; }
+
+  static NarrowType narrow_oop_value(narrowOop o) {
+    return static_cast<NarrowType>(o);
+  }
+
+  template<typename T>
+  static narrowOop narrow_oop_cast(T i) {
+    static_assert(std::is_integral<T>::value, "precondition");
+    static_assert(sizeof(T) >= sizeof(narrowOop), "precondition");
+    // Shift by 32 is UB if size in bits of i is 32, e.g. on 32bit platform.
+    assert(((i >> 16) >> 16) == 0, "narrowOop overflow");
+    return static_cast<narrowOop>(static_cast<NarrowType>(i));
```

Comment:
Should we have an assert that further restricts the range according to max heap size? 

Also, should we allow only unsigned types for T? I think using signed types here is prone to error and should be flagged.